src/react/utils.js

```diff
@@ -543,26 +543,41 @@ export function getComponentTypeFromRootValue(realm: Realm, value: Value): ECMAS
   }
 }
 
-// props should never have "ref" or "key" properties, as they're part of ReactElement
-// object instead. to ensure that we can give this hint, we create them and then
-// delete them, so their descriptor is left undefined. we use this knowledge later
-// to ensure that when dealing with creating ReactElements with partial config,
-// we don't have to bail out becuase "config" may or may not have "key" or/and "ref"
-export function deleteRefAndKeyFromProps(realm: Realm, props: ObjectValue | AbstractObjectValue): void {
-  setProperty(props, "ref", realm.intrinsics.undefined);
-  deleteProperty(props, "ref");
-  setProperty(props, "key", realm.intrinsics.undefined);
-  deleteProperty(props, "key");
-}
-
-export function objectHasNoPartialKeyAndRef(
-  realm: Realm,
-  object: ObjectValue | AbstractValue | AbstractObjectValue
-): boolean {
-  if (object instanceof AbstractValue) {
+export function flagPropsWithNoPartialKeyOrRef(realm: Realm, props: ObjectValue | AbstractObjectValue): void {
+  realm.react.propsWithNoPartialKeyOrRef.add(props);
+}
+
+export function hasNoPartialKeyOrRef(realm: Realm, props: ObjectValue | AbstractObjectValue): boolean {
+  if (realm.react.propsWithNoPartialKeyOrRef.has(props)) {
+    return true;
+  }
+  if (props instanceof ObjectValue && !props.isPartialObject()) {
+    realm.react.propsWithNoPartialKeyOrRef.add(props);
     return true;
   }
-  return !(Get(realm, object, "key") instanceof AbstractValue || Get(realm, object, "ref") instanceof AbstractValue);
+  if (props instanceof AbstractObjectValue) {
+    if (props.values.isTop()) {
+      return false;
+    }
+    let elements = props.values.getElements();
+    if (elements.size === 1) {
+      props = Array.from(elements)[0];
+    } else {
+      for (let element of elements) {
+        let wasSafe = hasNoPartialKeyOrRef(realm, element);
+        if (!wasSafe) {
+          return false;
+        }
+      }
+      realm.react.propsWithNoPartialKeyOrRef.add(props);
```

Comment:
It can't ever get them though, we're creating a ReactElement here and making props from the config. The props are immutable and can't change, so it can't get new properties that are partial.

Reply:
OK, I missed the fact that `hasNoPartialKeyOrRef` assumes its argument is a React `props` object and not any arbitrary object. Can you rename this method to make this assumption clear? e.g. `hasNoPartialKeyOrRefOnReactPropsObject`.

Reply:
Maybe invariant if the argument is not final would be helpful too.

Reply:
I also don't see this rule being consistently followed.
For example: https://github.com/facebook/prepack/pull/1996/files#diff-f17c375ba23f7377a789f6f90f5ea0cdR248

As far as I can tell you're passing `spreadValue` which is not guaranteed to be a "React props" immutable object. It could anything, couldn't it?

<img width="519" alt="screen shot 2018-05-25 at 6 49 47 pm" src="https://user-images.githubusercontent.com/810438/40558839-855850ae-604c-11e8-8311-1a14f8f6682e.png">

Reply:
Ignore my last comment, it was incorrect. I replied on my phone and didn't see the context properly (github is terrible on mobile).